src/async_fuse/memfs/kv_engine/etcd_impl.rs

```diff
@@ -1,18 +1,396 @@
+
+            tokio::select! {
+                _ = interval.tick() => {
+                    // Try to send a keep alive request
+                    match keeper.keep_alive().await {
+                        Ok(o) => {
+                            debug!("keep alive lease, lease_id={lease_id}, {o:?}");
+                        }
+                        Err(e) => {
+                            error!("failed to keep alive lease, error={e:?}");
+                            break;
+                        }
+                    }
+
+                    // Try to parse
+                    match lease_keep_alive_stream.message().await {
+                        Ok(Some(lease_alive_response)) => {
+                            if lease_alive_response.ttl() == 0 {
+                                error!("lease keep alive stream closed, because ttl is 0");
+                                self.close();
+                                break;
+                            }
+                            continue
+                        }
+                        Ok(None) => {
+                            error!("lease keep alive stream closed, because the sender is closed");
+                            self.close();
+                            break;
+                        }
+                        Err(e) => {
+                            error!("failed to keep alive lease, error={e:?}");
+                            self.close();
+                            break;
+                        }
+                    }
+                }
+                _ = close_rx.recv() => {
+                    info!("close the keep alive session, lease_id={lease_id}");
+                    self.close();
+                    break;
+                }
+                () = token.cancelled() => {
+                    debug!("lease keep alive stream closed by token canceled");
+                    self.close();
+                    break;
+                }
+            }
+        }
+
+        if self.is_closed() {
+            debug!("lease keep alive stream closed by is_closed flag");
+
+            // revoke the lease
+            revoke_lease!(client, lease_id);
+        }
+    }
+}
+
+/// The kvengine watch stream for etcd
+/// Try to receice message from etcd, and return key and value for each event,
+/// Wrap the etcd watch stream to support the `KVEngine` trait.
+#[derive(Debug)]
+pub struct KVEngineWatchStream {
+    /// The etcd client, used to cancel the watch stream
+    watcher: etcd_client::Watcher,
+    /// The etcd watch stream
+    watch_stream: etcd_client::WatchStream,
+}
+
+impl KVEngineWatchStream {
+    /// Create a new watch stream
+    #[must_use]
+    pub fn new(watcher: etcd_client::Watcher, watch_stream: etcd_client::WatchStream) -> Self {
+        Self {
+            watcher,
+            watch_stream,
+        }
+    }
+
+    /// Cancel the watch stream
+    #[inline]
+    pub async fn cancel(&mut self) -> DatenLordResult<()> {
```

Comment:
Better change `cancel`  to `async fn close(self)`, which will consume the obj ownership

Reply:
`watcher.cancel()` from `etcd_client` need a mutable value.

Reply:
Use `async fn close(mut self)` then, you should consume the ownership

Reply:
Ok, I have updated it.